src/chainstate/stacks/boot/exit-at-rc.clar

```diff
@@ -0,0 +1,199 @@
+;; This function enforces bounds on the vote (can't be above/below specific values).
+;; If a vote is accepted, the voter can only re-vote when they stack again.
+(define-public (vote-for-exit-rc (proposed-exit-rc uint))
+    (let (
+        (stacker-info (unwrap! (contract-call? .pox get-stacker-info tx-sender) (err ERR_VOTER_NOT_STACKING)))
+        (amount-stacked (get amount-ustx stacker-info))
+        (stacking-expiration (+ (get lock-period stacker-info) (get first-reward-cycle stacker-info)))
+        (current-reward-cycle (unwrap-panic (current-pox-reward-cycle)))
+        (proposed-rc-list (list proposed-exit-rc proposed-exit-rc proposed-exit-rc proposed-exit-rc proposed-exit-rc proposed-exit-rc proposed-exit-rc proposed-exit-rc proposed-exit-rc proposed-exit-rc proposed-exit-rc proposed-exit-rc))
+        (amount-stacked-list (list amount-stacked amount-stacked amount-stacked amount-stacked amount-stacked amount-stacked amount-stacked amount-stacked amount-stacked amount-stacked amount-stacked amount-stacked))
+        (lock-period (get lock-period stacker-info))
+        (lock-period-list (list lock-period lock-period lock-period lock-period lock-period lock-period lock-period lock-period lock-period lock-period lock-period lock-period))
+        (first-reward-cycle (get first-reward-cycle stacker-info))
+        (first-reward-cycle-list (list first-reward-cycle first-reward-cycle first-reward-cycle first-reward-cycle first-reward-cycle first-reward-cycle first-reward-cycle first-reward-cycle first-reward-cycle first-reward-cycle first-reward-cycle first-reward-cycle))
+        (list-indexes (list u0 u1 u2 u3 u4 u5 u6 u7 u8 u9 u10 u11))
+        ;; a list of the reward cycles the voter's vote is valid for, corresponding to the cycles their stx is locked for
+        (voting-reward-cycles (map get-voting-reward-cycles list-indexes lock-period-list first-reward-cycle-list ))
+    )
+        ;; Check that the amount stacked is positive
+        (asserts! (> amount-stacked u0) (err ERR_AMOUNT_NOT_POSITIVE))
+
+        ;; Check that the proposed exit rc is allowable
+        (asserts! (>= proposed-exit-rc (var-get absolute-minimum-exit-rc)) (err ERR_INVALID_PROPOSED_RC))
+        (asserts! (>= proposed-exit-rc (+ current-reward-cycle MINIMUM_RC_BUFFER_FROM_PRESENT)) (err ERR_INVALID_PROPOSED_RC))
+        (asserts! (<= proposed-exit-rc (+ current-reward-cycle MAXIMUM_RC_BUFFER_FROM_PRESENT)) (err ERR_INVALID_PROPOSED_RC))
+
+        ;; Check that the voter does not have an outstanding vote for this reward cycle
+        (match (map-get? voter-state {address: tx-sender})
+            voter-info (asserts! (>= current-reward-cycle (get expiration-reward-cycle voter-info))
+                (err ERR_PREVIOUS_VOTE_VALID))
+            ;; no existing state
+            true
+        )
+
+        ;; Check that the caller is allowed
+        (asserts! (is-eq tx-sender contract-caller) (err ERR_UNAUTHORIZED_CALLER))
```

Comment:
This effectively prevents stacking pools from voting, which I don't think we want?  Do we instead want to check that `contract-caller` is authorized explicitly by `tx-sender` to vote on its behalf, like we do in `.pox`?  Ideally the authorization state from `.pox` would be available to this code so the authorization could just carry over, but unfortunately it's private.

Reply:
hmm.. I wonder if it's a possibility to expose this in pox v2. In that release of the blockchain, we could add the exit contract as a boot contract, and use that authorization function. 

For this next breaking change though, I can add custom authorization logic to the exit contract.

Reply:
I don't think this prevents pools from voting. A pool can still vote by using `(as-contract ..)` to invoke this contract's methods. The bigger issue for pools' participation is that they won't be able to invoke these methods "as the contract" because the pool contracts are already all published, and they don't have methods for invoking the exit at contract.